src/mpd.c

```diff
@@ -3193,6 +3212,291 @@ mpd_command_update(struct evbuffer *evbuf, int argc, char **argv, char **errmsg,
   return 0;
 }
 
+static void
+mpd_add_idle_events(short event_mask);
+
+enum mpd_sticker_cmd_id {
+  MPD_STICKER_GET,
+  MPD_STICKER_SET,
+  MPD_STICKER_DELETE,
+  MPD_STICKER_LIST,
+  MPD_STICKER_FIND,
+  MPD_STICKER_INVALID
+};
+
+struct mpd_sticker_cmd {
+  const char *cmd;
+  enum mpd_sticker_cmd_id id;
+  int need_args;
+  int want_dir;
+};
+
+static struct mpd_sticker_cmd mpd_sticker_cmd[] = {
+  { "get", MPD_STICKER_GET, 5, 0 },
+  { "set", MPD_STICKER_SET, 6, 0 },
+  { "delete", MPD_STICKER_DELETE, 5, 0 },
+  { "list", MPD_STICKER_LIST, 4, 0 },
+  { "find", MPD_STICKER_FIND, 5, 1 },
+  { NULL, MPD_STICKER_INVALID, 0, 0 },
+};
+
+/*
+ * Command handler function for 'sticker'
+ *
+ *   sticker get "noth here" rating
+ *   ACK [2@0] {sticker} unknown sticker domain
+ *
+ *   sticker get song "Al Cohn & Shorty Rogers/East Coast - West Coast Scene/04 Shorty Rogers - Cool Sunshine.flac" rating
+ *   ACK [50@0] {sticker} no such sticker
+ *
+ *   sticker get song "Al Cohn & Shorty Rogers/East Coast - West Coast Scene/03 Al Cohn - Serenade For Kathy.flac" rating
+ *   sticker: rating=8
+ *   OK
+ *
+ * From cantata:
+ *   sticker set song "file:/srv/music/VA/The Electro Swing Revolution Vol 3 1 - Hop, Hop, Hop/13 Mr. Hotcut - You Are.mp3" rating "6"
+ *   OK
+ */
+static int
+mpd_command_sticker(struct evbuffer *evbuf, int argc, char **argv, char **errmsg, struct mpd_client_ctx *ctx)
+{
+  struct mpd_sticker_cmd *cmd_param;
+  const char *cmd;
+  const char *path;
+  const char *name;
+  int name_is_rating = 0;
+  int set_rating = 0;
+  int rating = -1;
+  int file_id;
+  struct media_file_info *mfi = NULL;
+  int i;
+  int ret;
+
+  if (argc < 4)
+    {
+      *errmsg = safe_asprintf("not enough arguments");
+      return ACK_ERROR_ARG;
+    }
+
+  if (strcmp(argv[2], "song") != 0)
+    {
+      *errmsg = safe_asprintf("unknown sticker domain");
+      return ACK_ERROR_ARG;
+    }
+
+  cmd = argv[1];
+  for (i=0; i<(sizeof(mpd_sticker_cmd) / sizeof(struct mpd_sticker_cmd)); ++i)
+    {
+      cmd_param = &mpd_sticker_cmd[i];
+      if (cmd_param->cmd && strcmp(cmd, cmd_param->cmd) == 0)
+	break;
+    }
+  if (cmd_param->id == MPD_STICKER_INVALID)
+    {
+      *errmsg = safe_asprintf("bad request");
+      return ACK_ERROR_ARG;
+    }
+  if (argc < cmd_param->need_args)
+    {
+      *errmsg = safe_asprintf("not enough arguments");
+      return ACK_ERROR_ARG;
+    }
+
+  if (argc > 4)
+    {
+      name = argv[4];
+      if (ctx->rating_name && ctx->rating_name[0] != '\0' && strcmp(name, ctx->rating_name) == 0)
+	name_is_rating = 1;
+    }
+
+  path = argv[3];
+  if (strncmp(path, "file:", strlen("file:")) == 0)
+    path += strlen("file:");
+
+  if (!cmd_param->want_dir && (name_is_rating || cmd_param->id == MPD_STICKER_LIST))
+    {
+      file_id = db_file_id_bypath(path);
```

Comment:
`db_file_fetch_byvirtualpath` will give you the media_file_info for a virtual path. Modifying the path only works for local files (and not for streams or spotify songs).